### PR TITLE
Allow configuring a timeout for `Endpoint` selection in `EndpointGroup`

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

A `DynamicEndpointGroup` lazily resolves `Endpoint`s even after a
request is started. Armeria clients use a connection timeout to wait for
the late `Endpoint`s.
https://github.com/line/armeria/blob/cc1656e6c9f7ba6b04d115e0298e90b8cf7c0971/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java#L320
The default connection timeout is 3.2 seconds which might be enough for a
single `EndpointGroup`.

However, `HealthCheckedEndpointGroup` wraps an `EndpointGroup` that
could be a `DynamicEndpointGroup` as well. Let's say the delegate
`EndpointGroup` is a DNS `EndpointGroup`. If a request is started, a DNS
resolution and HTTP health check should be preceded.  If a server
starts, it could take more resources and 3.2 seconds that can not make
sense to a complex `EndpointGroup` structure.

Modifications:

- Add `selectionTimeoutMillis()` to `EndpointGroup` so that a
  `EndpointSelector` uses the value to schedule a timeout.
- Add `selectionTimeout{Millis}(long)` to
  `AbstractDynamicEndpointGroupBuilder` in order to override the
  timeout.
  - Set 0 to the default selection timeout for static endpoints.
  - Set the default connection timeout to the default selection timeout
    for `DynamicEndpointGrouop`.
  - Set the default response timeout to the default selection timeout
    for `HealthCheckedEndpointGroup`, `ZooKeeperEndpointGroup`,
    `EurekaEndpointGroup` and `ConsulDnsEndpointGroup`.
  - `HealthCheckedEndpointGroup` adds its `selectionTimeout` and
    delegate's `selectionTimeout`.
- Raise `EndpointSelectionTimeoutException` if `EndpointSelctor` fails
  to select an `Endpoint` within a timeout.
- Make `ZooKeeperEndpointGroupBuilder` and `EurekaEndpointGroupBuilder`
  implment `DynamicEndpointGroupSetters`.
- Deprecate) `AbstractEndpointSelector.select(ctx, executor, timeout)`
  in favor of `AbstractEndpointSelector.select(ctx, executor)`

Result:

- You can now configure a timeout for an `Endpoint` selection of an
  `DynamicEndpointGroup`.
  ```java
  DnsAddressEndpointGroup delegate = DnsAddressEndpointGroup.of(...);
  HealthCheckedEndpointGroup healthGroup =
    HealthCheckedEndpointGroup
      .builder(delegate, "/health")
      .selectionTimeout(Duration.ofSeconds(10))  // 👈👈👈
      .build();
  ```

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java:303]: nit; what do you think of adding an explanation that setting this will set both `selectionTimeoutMillis` and `initialSelectionTimeoutMillis`

**Review by jrhee17 — COMMENTED** [review]
Looks mostly good! Left some minor questions

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:39]: Sorry for the late response.

> I think the default for DnsEndpointGroupBuilder will be connectTimeout instead of responseTimeout. I wanted to check if this is your intention.

Currently, the default select timeout for `DnsEndpointGroup` is the default connection timeout. 
https://github.com/line/armeria/blob/af45d0b2ae59c1a6c5a543257871b12dca92edfe/core/src/test/java/com/linecorp/armeria/client/endpoint/SelectionTimeoutTest.java#L118
DNS servers usually return a response quickly because they cache DNS records. Therefore, 3.2 seconds may be reasonable.

By the way, I checked the default timeout for a DNS query. It is 5 seconds that can be a more sensible default timeout for `DnsEndpointGroup/

> what do you think of just removing the empty constructor?

As this class is a public API, that could cause a breaking change.

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:39]: > As this class is a public API, that could cause a breaking change.

AbstractDynamicEndpointGroupBuilder is part of unstable APIs. Will remove the default constructor.

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java]: It sounds reasonable to use the same timeout for `DnsEndpointGroup` and our DNS resolver.

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractDynamicEndpointGroupBuilder.java:80]: maybe `final`?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java:149]: Could you also document what happens on the initial selection attempt? Will it wait indefinitely?

File: core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/dns/DnsEndpointGroupBuilder.java:127]:
```suggestion
     * If unspecified, the default DNS query timeout (5000 ms) is used by default.
```

File: core/src/main/java/com/linecorp/armeria/internal/client/dns/DnsUtil.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/client/dns/DnsUtil.java:126]: Can we use a static final field instead of a method so we can use `{@value ..}` when mentioning the default?

File: core/src/test/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroupBuilderTest.java
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroupBuilderTest.java:41]: 😱

File: core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/AbstractHealthCheckedEndpointGroupBuilder.java:264]: Ditto

**Review by jrhee17 — APPROVED** [review]
Looks good to me once @trustin 's comments are addressed 👍  Thanks @ikhoon ! 🙇 🚀 👍

File: core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java:86]: nit: Would it be better to use `-1` to indicate `StaticEndpointGroup`?

**Review by minwoox — APPROVED** [review]
Thanks a lot!
LGTM once @trustin's comments are addressed. 😉

File: core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroup.java:154]: Note: Will change the `{@code` to `{@link` after #4269 is merged.

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupSetters.java
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupSetters.java:46]: `selectionTimeout` or `selectTimeout`? I feel like `selectionTimeout` is correct, but we have `connectTimeout` 🤔

**Review by trustin — COMMENTED** [review]
Just one comment. Looks great otherwise!

File: core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupSetters.java
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroupSetters.java:46]: Probably `selectionTimeout` is better here because we have `EndpointSelectionTimeoutException`.

**Review by trustin — APPROVED** [review]
Nice work, @ikhoon 🙇

**Pull request merged**
